src/gurt/dtm.c

```diff
@@ -0,0 +1,425 @@
+				     type->dt_count, type->dt_reg.dr_name);
+			active_descriptors = true;
+		}
+		D_MUTEX_UNLOCK(&type->dt_lock);
+	}
+	D_MUTEX_UNLOCK(&dtm->dtm_lock);
+	return active_descriptors;
+}
+
+/* Create a single new object
+ *
+ * Returns a pointer to the object or NULL if allocation fails.
+ */
+static void *
+create(struct d_dtm_type *type)
+{
+	void *ptr;
+
+	D_ALLOC(ptr, type->dt_reg.dr_size);
+	if (!ptr)
+		return NULL;
+
+	type->dt_init_count++;
+	if (type->dt_reg.dr_init)
+		type->dt_reg.dr_init(ptr, type->dt_dtm->dtm_arg);
+
+	if (type->dt_reg.dr_reset) {
+		if (!type->dt_reg.dr_reset(ptr)) {
+			D_TRACE_INFO(type, "entry %p failed reset", ptr);
+			D_FREE(ptr);
+			return NULL;
+		}
+	}
+	type->dt_count++;
+
+	return ptr;
+}
+
+/* Populate the free list
+ *
+ * Create objects and add them to the free list.  Creates one more object
+ * than is needed to ensure that if the HWM of no-restock calls is reached
+ * there will be no on-path allocations.
+ */
+static void
+create_many(struct d_dtm_type *type)
+{
+	while (type->dt_free_count < (type->dt_no_restock_hwm + 1)) {
+		void *ptr;
+		d_list_t *entry;
+
+		if (type->dt_reg.dr_max_free_desc != 0 &&
+		    type->dt_free_count >= type->dt_reg.dr_max_free_desc)
+			break;
+
+		ptr = create(type);
+		entry = ptr + type->dt_reg.dr_offset;
+
+		if (!ptr)
+			return;
+
+		d_list_add_tail(entry, &type->dt_free_list);
+		type->dt_free_count++;
+	}
+}
+
+/* Register a data type */
+struct d_dtm_type *
+d_dtm_register(struct d_dtm *dtm, struct d_dtm_reg *reg)
+{
+	struct d_dtm_type *type;
+	int rc;
+
+	if (!reg->dr_name)
+		return NULL;
+
+	D_ALLOC_PTR(type);
+	if (!type)
+		return NULL;
+
+	rc = D_MUTEX_INIT(&type->dt_lock, NULL);
+	if (rc != -DER_SUCCESS) {
+		D_FREE(type);
+		return NULL;
+	}
+
+	D_TRACE_UP(DB_ANY, type, dtm, reg->dr_name);
+
+	D_INIT_LIST_HEAD(&type->dt_free_list);
+	D_INIT_LIST_HEAD(&type->dt_pending_list);
+	type->dt_dtm = dtm;
+
+	type->dt_count = 0;
+	type->dt_reg = *reg;
+
+	create_many(type);
+
+	if (type->dt_free_count == 0) {
+		/* If create_many() failed to create any descriptors then
+		 * return failure, as it either means an early allocation
+		 * failure or a wider problem with the type itself.
+		 *
+		 * This works with the fault injection tests as precisely
+		 * one descriptor is created initially, if there were more
+		 * and one failed the error would not propagate and the
+		 * injected fault would be ignored - failing the specific
+		 * test.
+		 */
+		D_MUTEX_DESTROY(&type->dt_lock);
+		D_FREE(type);
+		return NULL;
+	}
+
+	D_MUTEX_LOCK(&dtm->dtm_lock);
+	d_list_add_tail(&type->dt_type_list, &dtm->dtm_list);
+	D_MUTEX_UNLOCK(&dtm->dtm_lock);
+
+	return type;
+}
+
+/* Acquire a new object.
+ *
+ * This is to be considered on the critical path so should be as lightweight
+ * as posslble.
+ */
+void *
+d_dtm_acquire(struct d_dtm_type *type)
+{
+	void		*ptr = NULL;
+	d_list_t	*entry;
+	bool		at_limit = false;
+
+	D_MUTEX_LOCK(&type->dt_lock);
```

Comment:
I don't think that this will work fine as a general functionality in the server stack. It would be better to have this allocation per xstream so that it can be lockless and called directly by VOS for instance.

Reply:
> I think we should add a xstream-aware allocator using the TLS (see under src/iosrv) to store private data.

I agree that this would want a per-xstream wrapper around it so successful acquire calls could be thread-local.  I assume Jeff is going to add that next?